gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+# UDP(DNS) Health Check for Kubernetes Deployment
+
+| Status        |  Proposed        |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [#3](https://github.com/coredns/rfc/pull/3) |
+| **Author(s)** | Jayesh Sharma (@[wjayesh](https://github.com/wjayesh))  |
+| **Sponsor**   |    |
+| **Updated**   | 2020-03-03                                           |
+| **Obsoletes** |  |
+
+## Objective
+
+CoreDNS is the cluster DNS server for Kubernetes and is very critical for the overall health of the Kubernetes cluster. It is important to monitor the health of CoreDNS itself and restarting or repairing any CoreDNS pods that are not behaving correctly.  
+ 
+While CoreDNS exposes a health check itself in the form of Kubernetes’ `livenessProbe`: 
+
+
+- The health check is not UDP (DNS) based. There have been cases where the health port is accessible (for TCP) but CoreDNS itself isn't (UDP). This protocol difference means that CoreDNS is unhealthy from a cluster standpoint, but the control plane can't see this.  
+- The existing health check is also launched locally (the `kubelet` uses the `livenessProbe`) and the situation could be different for pods accessing it remotely. 
+ 
+## Motivation and Scope
+
+
+This project idea aims to get around limitations on Kubernetes’ health check and build an application that: 
+
+
+- Checks CoreDNS health externally through UDP (DNS), from a remote Golang application. 
+- Restart CoreDNS pods by interacting with Kubernetes API through the Golang application, if the response from the cluster and pod IPs is unsatisfactory.  
```

Comment:
@WJayesh , here, e.g. "The application restarts CoreDNS pods via Kubernetes API when the previously executed validations fail." 

> IPs is unsatisfactory

Here, you have a valid point. Please explain what the "unsatisfactory" means? Do I restart the pods when I see 1 error, 10 errors, 50% errors. It is worth discussing your logic here.

Reply:
Can the pods be restarted at every error? This is because a short outage might also be costly as all services rely on it. I'm not confident about this; what do you suggest? 
Keeping the first point in place, can a condition be added that if they are shutting down in quick succession, the situation might be out of control for the pod (as in Zalando's case) and then send a notification to the engineers?

Reply:
@WJayesh, make this part “tunable”, I.e. configuration options. It could be a percentage of failure ... basically thing through the proposal.